### PR TITLE
[Requirements] 12. Save GeoJSON data locally

### PR DESCRIPTION
**Pull request by OmarMuhammedAli**
# Save GeoJSON data locally

## Pull Request Type

🏆 **Enhancements** - This is an implementation of the 12th requirement of the PRD which enables the user to save GeoJSON data locally to `.geojson` and `.csv` formats. Other formats may be added in the future.

## Feedback required over

- A quick pair of  :eyes: on the code
- Discussion on the technical approach
- critique on design
- A review of copy

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -71,3 +72,6 @@ toml = "==0.10.2"
 
 [requires]
 python_version = "3.8"
+
+[pipenv]
+allow_prereleases = true
```
Comment (Rubix982) [inline Pipfile:77]: What does this do? :O

**Review by Rubix982 — APPROVED** [review]
All looks great! LGTM.

**Review by cbeddow — APPROVED** [review]
Looks good, all clear

**Comment by Rubix982** [comment]
@OmarMuhammedAli should I merge this will you be adding new commits?

**Comment by OmarMuhammedAli** [comment]
> @OmarMuhammedAli should I merge this will you be adding new commits?

I'm still implementing some of the changes that @cbeddow mentioned, so I'll be adding new commits

**Comment by OmarMuhammedAli** [comment]
@Rubix982 @cbeddow, I've introduced several changes in both controllers. Can you re-review this PR?

**Comment by Rubix982** [comment]
@OmarMuhammedAli sure!

**Review by Rubix982 — APPROVED** [review]
LGTM!

File: mapillary/controller/save.py [hunk 2]
```diff
@@ -13,38 +13,125 @@
...
+    # Ensure that the geojson is a dictionary
+    if isinstance(data, str):
+        data = json.loads(data)
+    
+    # Flatten the geojson
+    features = flatten_geojson(data)
+
+    try:
+        # Set the file name according to the given value. Default is "geojson.csv"
+        file_name = file_name if file_name is not None else "geojson.csv"
```
Comment (cbeddow) [inline mapillary/controller/save.py:51]: maybe just call this mapillary.csv, or maybe even `mapillary_{timestamp_milliseconds}.csv'. Not sure if that's a good idea or not, one would overwrite the same file, one would make a new file each time, both have advantages, but of course the user can just set their own file name and path too

File: mapillary/controller/save.py [hunk 3]
```diff
@@ -13,38 +13,125 @@
...
+    :return: Npne
+    :rtype: None
+    """
+    # Ensure that the geojson is a dictionary
+    if isinstance(data, str):
+        data = json.loads(data)
+    
+    try:
+        # Set the file name according to the given value. Default is "geojson.geojson"
+        file_name = file_name if file_name is not None else "geojson.geojson"
```
Comment (cbeddow) [inline mapillary/controller/save.py:128]: same here, at the least I think `mapillary.geojson` so they know it's from mapillary if losing track later (I do this a lot with geojson.io, they call it map.geojson and I forget where the data came fro) or maybe a timestamp to differentiate and prevent overwrite of the same one.

File: mapillary/mapillary.py [hunk 4]
```diff
@@ -850,52 +856,64 @@ def image_from_key(key: int, fields: list = []) -> str:
...
-        # TODO: the function works
-    """
+    :param format: The format to save the data as. Defaults to 'geojson'
+    :type format: str
 
-    # TODO: This functions needs implementation
+    Note: Allowed file format values at the moment are:
+        - geojson
+        - CSV
+    more file format values can be added and supported in the future
```
Comment (cbeddow) [inline mapillary/mapillary.py:883]: This is a good candidate for other first issues too

File: mapillary/mapillary.py [hunk 5]
```diff
@@ -850,52 +856,64 @@ def image_from_key(key: int, fields: list = []) -> str:
...
-        # TODO: the function works
-    """
-
-    # TODO: This functions needs implementation
+        >>> import mapillary as mly
+        >>> mly.set_access_token('MLY|XXX')
+        >>> mly.save(
+        ...     geojson_data=geojson_data,
+        ...     file_path=os.path.dirname(os.path.realpath(__file__)),
+        ...     file_name='test_geojson',
```
Comment (cbeddow) [inline mapillary/mapillary.py:895]: Maybe good to have a small check: if the user tries to put ".geojson" or ".csv" in the filename, then would save geojson.geojson.geojson. I've always wondered if that's worth throwing a warning or error to say the extension should not be in the file_name variable, but maybe it's okay to just leave it (the file will still open fine)

Comment (cbeddow) [reply mapillary/mapillary.py:895]: Maybe the thing is to just prevent it from having a "." in the filename but still, that is more for the appearance than the function, so I wonder what you think

**Review by cbeddow — APPROVED** [review]
Great work, some thoughts and ideas in the comments

File: mapillary/controller/save.py (hunk 2 above)
Comment (Rubix982) [reply mapillary/controller/save.py:51]: We can change this to f'mapillary_{int(datetime.datetime.now().timestamp()) * 1000}.csv' which gets the current UNIX timestamp.

File: mapillary/mapillary.py (hunk 4 above)
Comment (Rubix982) [reply mapillary/mapillary.py:883]: @cbeddow do you have some file formats in mind? I can make some quick issues with them.

File: mapillary/mapillary.py (hunk 5 above)
Comment (Rubix982) [reply mapillary/mapillary.py:895]: @OmarMuhammedAli maybe we should put a check on what a user can name their files? Only ASCII characters, and what Chris suggested.

File: mapillary/controller/save.py (hunk 2 above)
Comment (OmarMuhammedAli) [reply mapillary/controller/save.py:51]: Done and will be updating in the upcoming commit!

**Comment by OmarMuhammedAli** [comment]
Applied suggested changes. @cbeddow and @Rubix982, what do you think?

File: mapillary/mapillary.py (hunk 5 above)
Comment (OmarMuhammedAli) [reply mapillary/mapillary.py:895]: Done

**Comment by Rubix982** [comment]
Seems great, @OmarMuhammedAli! I think it can be merged.

**Comment by cbeddow** [comment]
Looks good! merging

**Pull request merged**